subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java

```diff
@@ -152,24 +149,31 @@ private JavaToolchain query(JavaToolchainSpec spec, boolean isFallback) {
         return downloadedToolchain;
     }
 
-    private InstallationLocation downloadToolchain(JavaToolchainSpec spec) {
-        final Optional<File> installation = installService.tryInstall(spec);
-        if (!installation.isPresent()) {
-            throw new NoToolchainAvailableException(spec, detectEnabled.getOrElse(true), downloadEnabled.getOrElse(true));
+    private void warnIfAutoProvisionedToolchainUsedWithoutRepositoryDefinitions(JavaToolchainInstantiationResult detectedToolchain) {
+        boolean autoDetectedToolchain = detectedToolchain.getJavaHome().isAutoProvisioned();
+        if (autoDetectedToolchain && installService.isAutoDownloadEnabled() && !installService.hasConfiguredToolchainRepositories()) {
+            LOGGER.warn("Build is using a toolchain installed via auto-provisioning (see " +
```

Comment:
I believe we should be using some form of deprecation logging here.
So that:
* It only gets printed once
* It is surfaced in build scans